machine_learning/k_means_clustering.c

```diff
@@ -0,0 +1,272 @@
+  for (int i = 0; i < k; i++) {
+    printf("%g %g %g setrgbcolor\n", *(colors + 3 * i), *(colors + 3 * i + 1),
+           *(colors + 3 * i + 2));
+    for (j = 0; j < len; j++) {
+      if (pts[j].group != i) continue;
+      printf("%.3f %.3f c\n", (pts[j].x - cx) * scale + W / 2,
+             (pts[j].y - cy) * scale + H / 2);
+    }
+    printf("\n0 setgray %g %g s\n", (cent[i].x - cx) * scale + W / 2,
+           (cent[i].y - cy) * scale + H / 2);
+  }
+  printf("\n%%%%EOF");
+
+  // free accquired memory
+  free(colors);
+}
+
+/*! \fn calculateNearest
+ * Returns the index of centroid nearest to
+ * given observation
+ *
+ * @param o - observation
+ * @param clusters - array of cluster having centroids coordinates
+ * @param k - size of clusters array
+ */
+int calculateNearst(observation* o, cluster clusters[], int k) {
+  double minD = __DBL_MAX__;
+  double dist = 0;
+  int index = -1;
+  int i = 0;
+  for (; i < k; i++) {
+    /* Calculate Squared Distance*/
+    dist = (clusters[i].x - o->x) * (clusters[i].x - o->x) +
+           (clusters[i].y - o->y) * (clusters[i].y - o->y);
+    if (dist < minD) {
+      minD = dist;
+      index = i;
+    }
+  }
+  return index;
+}
+
+/*! \fn calculateCentroid
+ * Calculate centoid and assign it to the cluster variable
+ *
+ * @param observations - an array of observations whose centroid is calculated
+ * @param size - size of the observations array
+ * @param centroid - a reference to cluster object to store information of
+ * centroid
+ */
+void calculateCentroid(observation observations[], size_t size,
+                       cluster* centroid) {
+  size_t i = 0;
+  centroid->x = 0;
+  centroid->y = 0;
+  centroid->count = size;
+  for (; i < size; i++) {
+    centroid->x += observations[i].x;
+    centroid->y += observations[i].y;
+    observations[i].group = 0;
+  }
+  centroid->x /= centroid->count;
+  centroid->y /= centroid->count;
+}
+
+/*!  \fn kMeans
+ *   --K Means Algorithm--
+ * 1. Assign each observation to one of k groups
+ *    creating a random initial clustering
+ * 2. Find the centroid of observations for each
+ *    cluster to form new centroids
+ * 3. Find the centroid which is nearest for each
+ *    observation among the calculated centroids
+ * 4. Assign the observation to its nearest centroid
+ *    to create a new clustering.
+ * 5. Repeat step 2,3,4 until there is no change
+ *    the current clustering and is same as last
+ *    clustering.
+ * @param observations - an array of observations to cluster
+ * @param size - size of observations array
+ * @param k - no of clusters to be made
+ *
+ * @returns pointer to cluster object
+ */
+cluster* kMeans(observation observations[], size_t size, int k) {
+  cluster* clusters = NULL;
+  if (k <= 1) {
+    /*
+    If we have to cluster them only in one group
+    then calculate centroid of observations and
+    that will be a ingle cluster
+    */
+    clusters = (cluster*)malloc(sizeof(cluster));
```

Comment:
The error reported by `clang-tidy` is regarding uninitialized values and probably arise from this function. Using `calloc` or `memset` to initialize the structure elements to '0' might help. Same in the other if-else cases.